## PSMC

* Reminder to take notes, share info here: [Etherpad](https://pad.carpentries.org/2019-Oct-SMSC)
* For this part of the tutorial, we're going to work with the same sorted bam file for the reference genome, from which we already removed the duplicates
	+ The BAM file should be in your `variants` folder. If not, you can copy it from:  `/scratch/genomics/frandsenp/SMSC/pop_gen/variants/ref_siskin.sorted.bam.mdup.bam*`
	+ The reference scaffold should be in your `genome` folder. If not, you can get it from here: `/scratch/genomics/frandsenp/SMSC/pop_gen/genome/Contig86_pilon.fasta`
	+ Create a new directory in `pop_gen` for your `PSMC` results. Call it `psmc`.

### 1. Generate stats and average coverage from our file:

* Let's estimate the depth with `samtools`
	+ **module**: ```bioinformatics/samtools```
	+ **command**: 
	```samtools depth ../variants/ref_siskin.sorted.bam.mdup.bam > ../psmc/final_depth.out```
		```awk '{sum+=$3 ; n++} END {if(n>0) print sum/n;}' ../psmc/final_depth.out > ../psmc/averagecov.txt```
	
	+ Later you will need mindepth, which should be 1/3 of the `averagecov.txt` and maxdepth, which should be 2X `averagecov.txt`

### 2. Identify heterozygous sites using mpileup and bcftools, then export to a diploid `FASTQ` that can be converted into a special `PSMC` `FASTA`
* The two steps for this are outlined in the commands below:
	+ **module**: ```bioinformatics/samtools```
	+ **module**: ```bioinformatics/bcftools```
	+ **command**: 

	```
	samtools mpileup -C50 -uf ref.fa aln.bam | bcftools view -c - \
	| vcfutils.pl vcf2fq -d <min_depth> -D <max_depth> | gzip > diploid.fq.gz
	```
	

### 3. PSMC
* Create the `PSMC` `FASTA` file.
	+ **module**: ```bioinformatics/psmc```
	+ **command**: ```fq2psmcfa -q20 diploid.fq.gz > siskin.psmcfa```

* First, run a single `PSMC` run.
	+ **module**: `bioinformatics/psmc`
	+ **command**: ```psmc -N25 -t15 -r5 -p "4+25*2+4+6" -o siskin.psmc siskin.psmcfa```

* Now, we're going to run 100 `PSMC` bootstraps.
	+ **module**: `bioinformatics/psmc`
	+ **command**: ```parallel -j $NSLOTS "psmc -N25 -t15 -r5 -b -p "4+25*2+4+6" siskin.psmcfa -o round-{}.psmc" :::: <(seq 100)```

* And, finally, we can create the `PSMC` plot. 
```psmc_plot.pl -u 3.18e-09 -g 4.2 siskin_psmc siskin.psmc```