## PSMC

* Reminder to take notes, share info here: [Etherpad](https://pad.carpentries.org/2019-Oct-SMSC)
* For this part of the tutorial, we're going to work with the same sorted bam file for the reference genome, from which we already removed the duplicates
	+ The BAM file should be in your `variants` folder. If not, you can copy it from:  `/scratch/genomics/frandsenp/SMSC/pop_gen/variants/ref_siskin.sorted.bam.mdup.bam*`
	+ The reference scaffold should be in your `genome` folder. If not, you can get it from here: `/scratch/genomics/frandsenp/SMSC/pop_gen/genome/Contig86_pilon.fasta`
	+ Create a new directory in `pop_gen` for your `PSMC` results. Call it `psmc`.

### 1. Identify heterozygous sites using mpileup and bcftools, then export to a diploid `FASTQ` that can be converted into a special `PSMC` `FASTA`
* The two steps for this are outlined in the commands below. The `-d` parameter is for minimum depth, which should be about 1/3 of the average depth of coverage and `-D` is maximum depth, which should be about twice the depth of coverage. Submit this job from your `jobs` directory from a job file called `diploid_fastq.job`:
	+ **module**: ```bioinformatics/samtools```
	+ **module**: ```bioinformatics/bcftools```
	+ **command**: 

	```
	samtools mpileup -C50 -uf ../genome/Contig86_pilon.fasta ../variants/ref_siskin.sorted.bam.mdup.bam \
	| bcftools call -c - | 	vcfutils.pl vcf2fq -d 16 -D 120 \
	| gzip > ../psmc/diploid.fq.gz
	```
	

### 2. PSMC
* Create the `PSMC` `FASTA` file.
	+ **module**: ```bioinformatics/psmc```
	+ **command**: ```fq2psmcfa -q20 ../psmc/diploid.fq.gz > ../psmc/siskin.psmcfa```

* First, run a single `PSMC` run.
	+ **module**: `bioinformatics/psmc`
	+ **command**: ```psmc -N25 -t15 -r5 -p "4+25*2+4+6" -o ../psmc/siskin.psmc ../psmc/siskin.psmcfa```

* FYI, you can also run 100 `PSMC` bootstraps. However, this won't be very useful for us because our example is a single contig and the bootstraps will all be the same because there aren't additional contigs to resample.
	+ **module**: `bioinformatics/psmc`
	+ **module**: `bioinformatics/parallel`
	+ **command**: ```parallel -j $NSLOTS "psmc -N25 -t15 -r5 -b -p "4+25*2+4+6" ../psmc/siskin.psmcfa -o ../psmc/round-{}.psmc" :::: <(seq 100)```

* Finally, we can create the `PSMC` plot without bootstrapping. This doesn't take a long time so go ahead and open up a `qrsh` session and `cd` directly into the `pop_gen/psmc` directory.
```psmc_plot.pl -u 3.18e-09 -g 4.2 siskin_psmc siskin.psmc```

* If you have extra time, you can begin to run `PSMC` on the other resequencing samples.